File: api/path_copy_all_versions.go
package vaku

import (
	"errors"
	"sort"
)

var (
	// ErrPathCopyAllVersions when PathCopyAllVersions fails.
	ErrPathCopyAllVersions = errors.New("path copy all versions")
)

// PathCopyAllVersions copies all versions of a secret from source to destination.
// Only works on v2 kv engines for both source and destination.
// Deleted or destroyed versions are preserved by writing empty secrets.
func (c *Client) PathCopyAllVersions(src, dst string) error {
	// Validate source is KV v2
	_, srcVersion, err := c.mountInfo(src)
	if err != nil {
		return newWrapErr(src, ErrPathCopyAllVersions, err)
	}
	if srcVersion != mv2 {
		return newWrapErr(src, ErrPathCopyAllVersions, ErrMountVersion)
	}

	// Validate destination is KV v2
	_, dstVersion, err := c.dc.mountInfo(dst)
	if err != nil {
		return newWrapErr(dst, ErrPathCopyAllVersions, err)
	}
	if dstVersion != mv2 {
		return newWrapErr(dst, ErrPathCopyAllVersions, ErrMountVersion)
	}

	// Read metadata to get all versions
	meta, err := c.PathReadMeta(src)
	if err != nil {
		return newWrapErr(src, ErrPathCopyAllVersions, err)
	}

	// If no metadata or no versions, nothing to copy
	if meta == nil || len(meta.Versions) == 0 {
		return nil
	}

	// Get sorted list of version numbers
	versions := make([]int, 0, len(meta.Versions))
	for v := range meta.Versions {
		versions = append(versions, v)
	}
	sort.Ints(versions)

	// Copy each version in order
	for _, v := range versions {
		vMeta := meta.Versions[v]

		var data map[string]any
		if vMeta.Deleted || vMeta.Destroyed {
			// Write empty secret to preserve version position
			data = map[string]any{}
		} else {
			// Read the version data
			data, err = c.PathReadVersion(src, v)
			if err != nil {
				return newWrapErr(src, ErrPathCopyAllVersions, err)
			}
			// If data is nil (shouldn't happen for non-deleted), use empty map
			if data == nil {
				data = map[string]any{}
			}
		}

		// Write to destination
		err = c.dc.PathWrite(dst, data)
		if err != nil {
			return newWrapErr(dst, ErrPathCopyAllVersions, err)
		}
	}

	return nil
}

File: src/api/path_copy_all_versions.rs
//! `PathCopyAllVersions` — replicate every historical version of a KV v2 secret.

use crate::api::client::Client;
use crate::api::error::{Error, ErrorKind};
use crate::api::mounts::{mount_info, MountVersion};
use crate::api::secret::SecretVersionMeta;
use serde_json::Map;

impl Client {
    /// Copy every version (including deleted/destroyed placeholders) of `src`
    /// to `dst`. Both sides must be KV v2.
    pub async fn path_copy_all_versions(&self, src: &str, dst: &str) -> Result<(), Error> {
        validate_kv2(self.src().mount_provider.as_ref(), src)
            .await
            .map_err(|e| Error::wrap(src, ErrorKind::PathCopyAllVersions, Some(Box::new(e))))?;
        validate_kv2(self.dst().mount_provider.as_ref(), dst)
            .await
            .map_err(|e| Error::wrap(dst, ErrorKind::PathCopyAllVersions, Some(Box::new(e))))?;

        let meta = self
            .path_read_meta(src)
            .await
            .map_err(|e| Error::wrap(src, ErrorKind::PathCopyAllVersions, Some(Box::new(e))))?;
        let Some(meta) = meta else { return Ok(()) };
        if meta.versions.is_empty() {
            return Ok(());
        }

        // Sort by version number ascending so destination ordering is stable.
        let mut versions: Vec<i64> = meta.versions.keys().copied().collect();
        versions.sort_unstable();

        let dst_client = self.as_destination();
        for v in versions {
            let vmeta = meta.versions.get(&v).cloned().unwrap_or_default();
            let data = self.get_version_data(src, v, &vmeta).await?;
            dst_client
                .path_write(dst, Some(data))
                .await
                .map_err(|e| Error::wrap(dst, ErrorKind::PathCopyAllVersions, Some(Box::new(e))))?;
        }

        Ok(())
    }

    async fn get_version_data(
        &self,
        src: &str,
        version: i64,
        vmeta: &SecretVersionMeta,
    ) -> Result<Map<String, serde_json::Value>, Error> {
        if vmeta.deleted || vmeta.destroyed {
            return Ok(Map::new());
        }
        let data = self
            .path_read_version(src, version)
            .await
            .map_err(|e| Error::wrap(src, ErrorKind::PathCopyAllVersions, Some(Box::new(e))))?;
        Ok(data.unwrap_or_default())
    }
}

/// Returns `Ok(())` when `path` is on a KV v2 mount, otherwise [`MountVersion`]
/// or [`MountInfo`] errors.
pub(crate) async fn validate_kv2(
    provider: &dyn crate::api::mount_provider::MountProvider,
    path: &str,
) -> Result<(), Error> {
    let (_, version) = mount_info(provider, path).await?;
    if !matches!(version, MountVersion::Mv2) {
        return Err(Error::wrap("", ErrorKind::MountVersion, None));
    }
    Ok(())
}
